File: lc_soln/1448_count_good_nodes_in_binary_tree/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution
{
    public int GoodNodes(TreeNode root)
    {
        return Dfs(root, int.MinValue);
    }

    private int Dfs(TreeNode node, int maxVal)
    {
        if (node is null)
        {
            return 0;
        }
        maxVal = Math.Max(maxVal, node.val);
        int result = node.val >= maxVal ? 1 : 0;
        return result + Dfs(node.left, maxVal) + Dfs(node.right, maxVal);
    }
}

File: lc_soln/1448_count_good_nodes_in_binary_tree/solution.rs
struct Solution;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn good_nodes(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, max_val: i32) -> i32 {
            match node {
                None => 0,
                Some(n) => {
                    let n = n.borrow();
                    let result = if n.val >= max_val { 1 } else { 0 };
                    let new_max = max_val.max(n.val);
                    result + dfs(&n.left, new_max) + dfs(&n.right, new_max)
                }
            }
        }
        dfs(&root, i32::MIN)
    }
}

fn main() {}
